server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -396,24 +398,46 @@ private void deleteSynonymsSetObjects(String synonymSetId, ActionListener<BulkBy
-                // If nothing was deleted, synonym set did not exist
-                deleteObjectsListener.onFailure(new ResourceNotFoundException("Synonym set [" + synonymSetId + "] not found"));
-                return;
-            }
-            final List<BulkItemResponse.Failure> bulkFailures = bulkByScrollResponse.getBulkFailures();
-            if (bulkFailures.isEmpty() == false) {
-                deleteObjectsListener.onFailure(
-                    new ElasticsearchException(
-                        "Error deleting synonym set: "
-                            + bulkFailures.stream().map(BulkItemResponse.Failure::getMessage).collect(Collectors.joining("\n"))
+
+        // Checks reloading the analyzer to understand usage. We can iterate on this solution to target the specific analyzer
+        reloadAnalyzers(synonymSetId, listener.delegateFailure((reloadListener, reloadResult) -> {
+            Map<String, ReloadAnalyzersResponse.ReloadDetails> reloadDetails = reloadResult.reloadAnalyzersResponse.getReloadDetails();
+            if (reloadDetails.isEmpty() == false) {
+                Set<String> indices = reloadDetails.entrySet()
+                    .stream()
+                    .map(entry -> entry.getValue().getIndexName())
+                    .collect(Collectors.toSet());
+                reloadListener.onFailure(
+                    new IllegalArgumentException(
+                        "Synonym set ["
+                            + synonymSetId
+                            + "] cannot be deleted as it is used in the following indices: "
+                            + String.join(", ", indices)
                     )
                 );
                 return;
             }
-            reloadAnalyzers(synonymSetId, deleteObjectsListener, AcknowledgedResponse.of(true));
-        }));
+
+            deleteSynonymsSetObjects(synonymSetId, listener.delegateFailure((deleteObjectsListener, bulkByScrollResponse) -> {
+                if (bulkByScrollResponse.getDeleted() == 0) {
+                    // If nothing was deleted, synonym set did not exist
+                    deleteObjectsListener.onFailure(new ResourceNotFoundException("Synonym set [" + synonymSetId + "] not found"));
+                    return;
+                }
+                final List<BulkItemResponse.Failure> bulkFailures = bulkByScrollResponse.getBulkFailures();
+                if (bulkFailures.isEmpty() == false) {
+                    deleteObjectsListener.onFailure(
+                        new InvalidParameterException(
+                            "Error deleting synonym set: "
+                                + bulkFailures.stream().map(BulkItemResponse.Failure::getMessage).collect(Collectors.joining("\n"))
+                        )
+                    );
+                    return;
+                }
+                reloadAnalyzers(synonymSetId, deleteObjectsListener, AcknowledgedResponse.of(true));
```

Comment:
I think we can remove the reload analyzers operation after the synonym set has been deleted; we already know this is not being used, and other indices won't have to be reloaded.

Any reason to keep this? Otherwise, I'll change the delete response as well so it doesn't contain reload analyzer results